File: Common/src/main/java/at/petrak/hexcasting/api/casting/math/HexPattern.java
package at.petrak.hexcasting.api.casting.math;

import at.petrak.hexcasting.api.utils.HexUtils;
import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.phys.Vec2;

/**
 * Sequence of angles to define a pattern traced.
 */
public final class HexPattern {
    private final HexDir orientation;
    private final HexSignature signature;

    public HexPattern(HexDir orientation, HexSignature signature) {
        this.orientation = orientation;
        this.signature = signature;
    }

    public HexDir getOrientation() {
        return orientation;
    }

    public HexSignature getSignature() {
        return signature;
    }

    public Vec2 getCenter(float hexRadius) {
        return this.getCenter(hexRadius, HexCoord.getOrigin());
    }

    public Vec2 getCenter(float hexRadius, HexCoord origin) {
        Vec2 originPx = HexUtils.coordToPx(origin, hexRadius, Vec2.ZERO);
        List<Vec2> points = this.toLines(hexRadius, originPx);
        return HexUtils.findCenter(points);
    }

    /**
     * Convert a hex pattern into a sequence of straight linePoints spanning its points.
     */
    public List<Vec2> toLines(float hexSize, Vec2 origin) {
        return this.positions().stream().map(it -> HexUtils.coordToPx(it, hexSize, origin)).toList();
    }

    public List<HexCoord> positions() {
        return this.positions(HexCoord.getOrigin());
    }

    public List<HexCoord> positions(HexCoord start)  {
        ArrayList<HexCoord> out = new ArrayList<>(this.signature.sizeUpperBound() + 2);
        out.add(start);
        HexDir compass = this.orientation;
        HexCoord cursor = start;
        for (HexAngle a : this.signature) {
            cursor = cursor.plus(compass);
            out.add(cursor);
            compass = compass.times(a);
        }
        out.add(cursor.plus(compass));
        return out;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        HexPattern that = (HexPattern) o;
        return Objects.equals(signature, that.signature) && orientation == that.orientation;
    }

    @Override
    public int hashCode() {
        return Objects.hash(signature, orientation);
    }

    @Override
    public String toString() {
        return "HexPattern[" +
                "orientation=" + orientation +
                ", signature=" + signature +
                ']';
    }

    public String toChatString() {
        return "HexPattern[" + orientation + ", " + signature.toAnglesString() + ']';
    }

    public static final Codec<HexPattern> CODEC = RecordCodecBuilder.create(instance -> instance.group(
            HexDir.CODEC.fieldOf("orientation").forGetter(HexPattern::getOrientation),
            HexSignature.CODEC.fieldOf("signature").forGetter(HexPattern::getSignature)
    ).apply(instance, HexPattern::new));

    public static final StreamCodec<RegistryFriendlyByteBuf, HexPattern> STREAM_CODEC = StreamCodec.composite(
            HexDir.STREAM_CODEC, HexPattern::getOrientation,
            HexSignature.STREAM_CODEC, HexPattern::getSignature,
            HexPattern::new
    );

    public static HexPattern fromAngleString(String angleString, HexDir startDir) {
        return fromAngleString(angleString, startDir, true);
    }

    public static HexPattern fromAngleString(String angleString, HexDir startDir, boolean rejectOverlap) {
        HexSignature.Builder signature = new HexSignature.Builder(rejectOverlap);

        for (int i = 0; i < angleString.length(); i++) {
            char c = angleString.charAt(i);

            HexAngle angle = HexAngle.fromChar(c);
            if(angle == null) throw new IllegalArgumentException("Cannot match %c at idx %d to a direction".formatted(c, i));

            signature.addAngle(angle);
        }

        return new HexPattern(startDir, signature.build());
    }
}

File: Common/src/main/java/at/petrak/hexcasting/api/casting/math/HexPattern.kt
package at.petrak.hexcasting.api.casting.math

import at.petrak.hexcasting.api.utils.NBTBuilder
import at.petrak.hexcasting.api.utils.coordToPx
import at.petrak.hexcasting.api.utils.findCenter
import at.petrak.hexcasting.api.utils.getSafe
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.Tag
import net.minecraft.world.phys.Vec2

/**
 * Sequence of angles to define a pattern traced.
 */
data class HexPattern(val startDir: HexDir, val angles: MutableList<HexAngle> = arrayListOf()) {
    /**
     * @return True if it successfully appended, false if not.
     */
    fun tryAppendDir(newDir: HexDir): Boolean {
        // Two restrictions:
        // - No adding a pos/dir pair we previously added
        // - No backtracking
        val linesSeen = mutableSetOf<Pair<HexCoord, HexDir>>()

        var compass = this.startDir
        var cursor = HexCoord.Origin
        for (a in this.angles) {
            linesSeen.add(cursor to compass)
            // Line from here to there also blocks there to here
            linesSeen.add(cursor + compass to compass.rotatedBy(HexAngle.BACK))
            cursor += compass
            compass *= a
        }
        cursor += compass

        val potentialNewLine = cursor to newDir
        if (potentialNewLine in linesSeen) return false
        val nextAngle = newDir - compass
        if (nextAngle == HexAngle.BACK) return false

        this.angles.add(nextAngle)
        return true
    }

    @JvmOverloads
    fun positions(start: HexCoord = HexCoord.Origin): List<HexCoord> {
        val out: ArrayList<HexCoord> = ArrayList(this.angles.size + 2)
        out.add(start)
        var compass: HexDir = this.startDir
        var cursor = start
        for (a in this.angles) {
            cursor += compass
            out.add(cursor)
            compass *= a
        }
        out.add(cursor + compass)
        return out
    }

    fun directions(): List<HexDir> {
        val out = ArrayList<HexDir>(this.angles.size + 1)
        out.add(this.startDir)

        var compass: HexDir = this.startDir
        for (a in this.angles) {
            compass *= a
            out.add(compass)
        }
        return out
    }

    fun finalDir(): HexDir =
        this.angles.fold(this.startDir) { acc, angle -> acc * angle }


    fun serializeToNBT() = NBTBuilder {
        TAG_START_DIR %= byte(startDir.ordinal)
        TAG_ANGLES %= byteArray(angles.map(HexAngle::ordinal))
    }

    // Terrible shorthand method for easy matching
    fun anglesSignature(): String {
        return buildString {
            for (a in this@HexPattern.angles) {
                append(
                    when (a) {
                        HexAngle.FORWARD -> "w"
                        HexAngle.RIGHT -> "e"
                        HexAngle.RIGHT_BACK -> "d"
                        HexAngle.BACK -> "s"
                        HexAngle.LEFT_BACK -> "a"
                        HexAngle.LEFT -> "q"
                    }
                )
            }
        }
    }

    /**
     * Return the "center of mass" of the pattern.
     * Drawing the pattern with the returned vector as the origin will center the pattern around it.
     */
    @JvmOverloads
    fun getCenter(hexRadius: Float, origin: HexCoord = HexCoord.Origin): Vec2 {
        val originPx = coordToPx(origin, hexRadius, Vec2.ZERO)
        val points = this.toLines(hexRadius, originPx)
        return findCenter(points)
    }


    /**
     * Convert a hex pattern into a sequence of straight linePoints spanning its points.
     */
    fun toLines(hexSize: Float, origin: Vec2): List<Vec2> =
        this.positions().map { coordToPx(it, hexSize, origin) }

    fun sigsEqual(that: HexPattern) = this.angles == that.angles

    override fun toString(): String = buildString {
        append("HexPattern[")
        append(this@HexPattern.startDir)
        append(", ")
        append(this@HexPattern.anglesSignature())
        append("]")
    }

    companion object {
        const val TAG_START_DIR = "start_dir"
        const val TAG_ANGLES = "angles"

        @JvmStatic
        fun isPattern(tag: CompoundTag): Boolean {
            return tag.contains(TAG_START_DIR, Tag.TAG_ANY_NUMERIC.toInt())
                && tag.contains(TAG_ANGLES, Tag.TAG_BYTE_ARRAY.toInt())
        }

        @JvmStatic
        fun fromNBT(tag: CompoundTag): HexPattern {
            val startDir = HexDir.values().getSafe(tag.getByte(TAG_START_DIR))
            val angles = tag.getByteArray(TAG_ANGLES).map(HexAngle.values()::getSafe)
            return HexPattern(startDir, angles.toMutableList())
        }

        @JvmStatic
        fun fromAngles(signature: String, startDir: HexDir): HexPattern {
            val out = HexPattern(startDir)
            var compass = startDir

            for ((idx, c) in signature.withIndex()) {
                val angle = when (c) {
                    'w' -> HexAngle.FORWARD
                    'e' -> HexAngle.RIGHT
                    'd' -> HexAngle.RIGHT_BACK
                    // for completeness ...
                    's' -> HexAngle.BACK
                    'a' -> HexAngle.LEFT_BACK
                    'q' -> HexAngle.LEFT
                    else -> throw IllegalArgumentException("Cannot match $c at idx $idx to a direction")
                }
                compass *= angle
                val success = out.tryAppendDir(compass)
                if (!success) {
                    throw IllegalStateException("Adding the angle $c at index $idx made the pattern invalid by looping back on itself")
                }
            }
            return out
        }

    }
}
